### PR TITLE
python: Add --python-mode option

### PR DESCRIPTION
**Pull request by Jongy**
~~Based on https://github.com/Granulate/gprofiler/pull/108 and https://github.com/Granulate/gprofiler/pull/89, review only after.~~

## Description
Adds the `--python-mode` option:
* `auto` - try PyPerf, fallback to py-spy
* `pyperf` - PyPerf, and no fallback
* `pyspy` - py-spy w/o trying PyPerf
* `none` - like `--no-python`.

## Motivation and Context
Give tighter control over which profilers are run, exactly. It's possible to want to use PyPerf with no fallback, or to avoid trying PyPerf to begin with.

## How Has This Been Tested?
* Running with `auto` - PyPerf is selected, if I "hide" kernel headers then py-spy is selected
* Running with `auto` and killing PyPerf - `Python eBPF profiler failed (exit code: -15), falling back to py-spy`
* Running with `pyperf` and no kernel headers - no Python profiler is used
* Running with `pyperf` and PyPerf is killed - `Python eBPF profiler failed (exit code: -15)` and no pyspy next round.
* Running with `pyspy` just runs py-spy w/o PyPerf initialization
* Running with `none` runs no Python profiler at all.

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have updated the relevant documentation.
- [ ] I have added tests for new logic.

### CODE REVIEW

**Comment by michelhe** [comment]
Use `PythonProfiler` in `test_sanity.py`, the ebpf profiler is broken

**Comment by Jongy** [comment]
> Use `PythonProfiler` in `test_sanity.py`, the ebpf profiler is broken

Nope, it should remain `PythonEbpfProfiler` explicitly. The failure is just flakiness :( we happened to fall only on other syscalls and not on `getdents`. Which is really bad luck considering the frequency and duration...

**Comment by Jongy** [comment]
Ready for review

File: gprofiler/python.py [hunk 1]
```diff
@@ -118,6 +116,12 @@ def snapshot(self) -> ProcessToStackSampleCounters:
         return results
 
 
+class PythonEbpfError(CalledProcessError):
+    """ "
```
Comment (Jongy) [inline gprofiler/python.py:120]:
```suggestion
    """
```

File: gprofiler/python.py [hunk 2]
```diff
@@ -280,31 +276,67 @@ def stop(self):
...
 
+    def __init__(
+        self,
+        frequency: int,
+        duration: int,
+        stop_event: Event,
+        storage_dir: str,
+        python_mode: str,
+    ):
+        # not calling my c'tor - we're just a proxy class
```
Comment (Jongy) [inline gprofiler/python.py:293]: Is it relevant? I use `ProfilerInterface` here, not `ProfilerBase`

Comment (michelhe) [reply gprofiler/python.py:293]: Doesn't really matter but if someone adds a constructor later..

File: README.md [hunk 3]
```diff
@@ -47,13 +47,32 @@ The default profiling frequency is *11 hertz*. Using higher frequency will lead
...
+
+* `--no-java`: Disable profilers for Java.
+
+### Python profiling options
+* `--no-python`: Alias of `--python-mode none`.
+* `--python-mode`: Controls which profiler is used for Python.
+    * `auto` - (default) try with PyPerf, fall back to py-spy.
+    * `pyperf` - Use PyPerf with no py-spy fallback.
+    * `pyspy` - Use py-spy.
+    * `none` - Disable profilers for Python.
```
Comment (michelhe) [inline README.md:60]: worth to include explanation of pyspy vs PyPerf

File: gprofiler/main.py [hunk 4]
```diff
@@ -392,20 +390,31 @@ def parse_cmd_args():
         "--rotating-output", action="store_true", default=False, help="Keep only the last profile result"
     )
 
-    parser.add_argument(
+    java_options = parser.add_argument_group("Java")
+    java_options.add_argument(
```
Comment (michelhe) [inline gprofiler/main.py:394]: do the same for php & ruby

File: gprofiler/python.py [hunk 5]
```diff
@@ -245,18 +249,10 @@ def _dump(self) -> Path:
...
-            finally:
-                # in any case, we should stop PyPerf.
-                logger.warning("PyPerf dead/not responding, killing it")
-                process = self.process  # save it
-                self._terminate()
-                self._pyperf_error(process)
+            logger.warning("PyPerf dead/not responding, killing it")
+            process = self.process  # save it
+            self._terminate()
+            self._pyperf_error(process)
```
Comment (michelhe) [inline gprofiler/python.py:255]: Did you remove the try-finally for a reason ?

File: gprofiler/python.py (hunk 2 above)
Comment (Jongy) [reply gprofiler/python.py:293]: Will change to the interface

File: gprofiler/main.py (hunk 4 above)
Comment (Jongy) [reply gprofiler/main.py:394]: mhm

File: gprofiler/python.py (hunk 5 above)
Comment (Jongy) [reply gprofiler/python.py:255]: Yes - the swap logic is now done in `PythonProfiler.snapshot`, so the `try` part actually happens later, and the `finally` part can happen immediately.

File: README.md (hunk 3 above)
Comment (Jongy) [reply README.md:60]: k added a bit

File: gprofiler/python.py (hunk 2 above)
Comment (Jongy) [reply gprofiler/python.py:293]: Hmm I already use the interface. So comment is irrelevant.

**Review by michelhe — APPROVED** [review]
LGTM, Waiting for CI

**Review by michelhe — APPROVED** [review]

**Pull request merged**